### PR TITLE
Bite all cups after Global Settlement

### PR DESCRIPTION
**Pull request by KentonPrescott**
Among bringing the repo up to date (`pymaker`, adding `install.sh`, etc), I’ve added `bite_keeper/bite_all_keeper.py`, which semi-intelligently attempts to `bite` all `cups` that meet the following conditions:
* Hasn’t been bitten before (i.e. `cup.art != 0`)
* Holds ~`ink`~ `owe` above some X amount (currently set to `0`)

It’s been tested on Kovan, sends transactions synchronously, and utilizes the `IncreasingGasPrice` strategy (starting at 5 GWei), as defined in pymaker. Here are some Kovan transactions: 
* https://kovan.etherscan.io/tx/0x0399b2dc26757d65081d9f9972d4b59567758ef89e7c21b666030bde470a06af
* https://kovan.etherscan.io/tx/0xdd3e74b4a5b7803c5cd3801c9e396da545bfa643e0df063c4e0e1f2b05def1e4

Possible improvements:
- Consider changing `bite().transact()` -> `bite().transact_async()` in case performance is becomes a concern.
- Incorporate https://github.com/makerdao/pygasprice-client
- Gas cost optimizations

### CODE REVIEW

File: bite_keeper/bite_all_keeper.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            self.logger.info('Single Collateral Dai has been Caged')
+            self.logger.info('Starting to bite all cups in the tub contract')
+
+            for cup_id in range(self.tub.cupi()):
+                self.check_cup(cup_id+1)
+        else:
+            self.logger.info('Single Collateral Dai live')
+
+    def check_cup(self, cup_id):
+        cup = self.tub.cups(cup_id)
```
Comment (jparklev) [inline bite_keeper/bite_all_keeper.py]: question of ignorance, does `tub.cups` already do the converting to `bytes32` for you?

**Comment by levity** [comment]
Just commenting to note that the keeper should also call `tap.vent()`, either after each bite or just once after all of them, to burn all the PETH that was removed from CDPs.

this is necessary to keep `tub.per()`, the WETH/PETH ratio, up to date so that PETH holders get the right amount of WETH back when they `exit`.

File: bite_keeper/bite_all_keeper.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import logging
+import sys
+
+from web3 import Web3, HTTPProvider
+
+from pymaker import Address
+from pymaker.gas import FixedGasPrice, DefaultGasPrice, IncreasingGasPrice
+from pymaker.lifecycle import Lifecycle
+from pymaker.sai import Tub
+from pymaker.keys import register_keys
```
[the remark this answers is not in the document]
Comment (KentonPrescott) [reply bite_keeper/bite_all_keeper.py]: Once `--eth-key` is supplied, `register_keys()` handles the unlock of the keystore file. 
Older keeper repos (e.g. `arbitrage-keeper`) were developed without this facility and (my guess) required a local node that had access to unlocked accounts.

File: bite_keeper/bite_all_keeper.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            for cup_id in range(self.tub.cupi()):
+                self.check_cup(cup_id+1)
+        else:
+            self.logger.info('Single Collateral Dai live')
+
+    def check_cup(self, cup_id):
+        cup = self.tub.cups(cup_id)
+
+        # Bite cups with ink over a threshold that haven't been bitten before
+        if cup.ink > Wad.from_number(0) and cup.art > Wad(0):
```
[the remark this answers is not in the document]
Comment (KentonPrescott) [reply bite_keeper/bite_all_keeper.py]: `Wad(x)` assumes `x` is in units of wei.
`Wad.from_number(y)` assumes `y` is in units of eth.
I used the latter simply for ease of adjustment, if we decide on a `cup.ink` threshold.

I'm now realizing that I should have been consistent with `cup.art`. Will update now! Thanks for the catch :)

File: bite_keeper/bite_all_keeper.py (hunk 1 above)
Comment (KentonPrescott) [reply bite_keeper/bite_all_keeper.py]: [thankfully yes](https://github.com/makerdao/pymaker/blob/master/pymaker/sai.py#L442)

File: bite_keeper/bite_all_keeper.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KentonPrescott) [reply bite_keeper/bite_all_keeper.py]: @jparklev @gbalabasquer I changed to `cup.art != Wad.from_number(0)`, but should it be reverted to `cup.art > Wad.from_number(X)`? Is there a need to apply a threshold to the cup's debt, or are we only concerned about its collateral (`ink`)?

[the remark this answers is not in the document]
Comment (jparklev) [reply bite_keeper/bite_all_keeper.py]: > Is there a need to apply a threshold to the cup's debt, or are we only concerned about its collateral (ink)?

I think we're actually only concerned about the debt for `bite`. Looking at [this line](https://github.com/makerdao/sai/blob/00ccfd780295e86cee54f57b6cd0673b036ae370/src/tub.sol#L332) in tub, if `owe` is 0 (ie there is not debt), no `peth` is pushed to the `tap`

[the remark this answers is not in the document]
Comment (KentonPrescott) [reply bite_keeper/bite_all_keeper.py]: Ah, good call. Thanks for pointing that out. 
The most recent commit implements an off-chain calculation of `owe`, which will be used in the condition for `bite`ing. Changing the PR description to reflect the change.

**Comment by KentonPrescott** [comment]
To prevent scope creep, I’m going to merge this PR. 
Much thanks to @jparklev for pointing out `owe` as the more relevant metric to measure.
As @levity recommended, there are more improvements to be made before this keeper is finalized.

**Pull request merged**